### ISSUE TITLE
Proxy: Reverse-map set-cookie parameters to match proxy transformation

### ISSUE THREAD

**Issue by xkr47** [NitorCreations/nitor-backend#3]
if /foo is mapped to localhost:1234/bar using the Proxy handler and the remote server responds `Set-Cookie X=Y,path=/bar,domain=localhost:1234` then it should be mapped back to `path=/foo,domain=something`.

Corresponding Apache server configuration: https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#ProxyPassReverse

**Comment by psiniemi** [comment]
IMHO rewriting the url path on the fly is evil and is likely to bite you. These mappings should then potentially go to js, css and html content in addition to location headers and the like.

**Comment by xkr47** [comment]
Yeah it needs to be configurable. But it is based on real needs in current customer environments. One site is implemented so that all html, js css urls are relative to root, but the servlet container disallows mounting directly to root, so the site is served in /foo and cookies end up referring to that "wrong" path.. so we map it back to the root using the ProxyPass, ProxyPassReverse and ProxyPassReverseCookiePath settings.

**Comment by psiniemi** [comment]
A simpler solution would be to have a redirect /(.*) -> /foo/\1 and not touch the path. Anyway I guess this is not a huge feature and having feature parity here with basic apache is generally a good idea (even if the feature itself is bad).

**Comment by gmokki** [comment]
Your example of / -> /foo requires touching the path of the cookies the application sets. For example the if the application sets cookie to path `/foo` and proxy does not edit the cookie in the response. The browser will not send it in the next request because the app is in /app and not in /foo/app

**Comment by gmokki** [comment]
Or maybe I misunderstood and you mean the nice redirect hack proxy we did few years back that relies on the referers to figure out what is happening.

**Comment by psiniemi** [comment]
My simplification idea was that the app run in /foo, and to just redirect users there from the root. And the app would then agree on this context.

These tricks are going to be messy whatever you do and the app will have to be aware of how the browsers see it somehow.

I guess the point here is that since the feature is available in apache, lots of apps have been written to rely on this functionality and to replace apache you will have to provide the same.
